Fix srcset attribute setting order issues by aligning with spec

yoavweiss opened the pull request:
#### 1382c0d807cf9583aa6ca231ce6aae74c8473349
<pre>
Fix srcset attribute setting order issues by aligning with spec

<a href="https://bugs.webkit.org/show_bug.cgi?id=190031">https://bugs.webkit.org/show_bug.cgi?id=190031</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=276586">https://bugs.webkit.org/show_bug.cgi?id=276586</a>

Reviewed by NOBODY (OOPS!).

Currently the order of setting image attributes matter, as some
attribute setting processing model triggers an image load.
Therefore setting some of these before others can impact the asset being
downloaded (or if an asset is downloaded at all).

As specified, this should not happen, as image loading awaits
a microtask before being triggered.

This PR aligns the implementation with the spec, and fixes these related
issues.

* Source/WebCore/loader/ImageLoader.cpp:
 (WebCore::ImageLoader::updateFromElementIgnoringPreviousError): Await
 a microtask before calling updateFromElement.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-img-element/attribute-order-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-img-element/attribute-order.html: Added.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/68b54ebce773ef4dffd64b9cb58f4370848c32c6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73414 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52843 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26222 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77669 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24652 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75529 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/61972 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/626 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57684 "Found 3 new test failures: http/tests/media/media-blocked-by-willsendrequest.html imported/w3c/web-platform-tests/html/semantics/embedded-content/the-img-element/list-of-available-images-matching.https.html security/block-test.html (failure)") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16133 "Found 1 new test failure: fast/events/onerror-img-after-gc.html (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76481 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47762 "Found 1 new test failure: fast/events/onerror-img-after-gc.html (failure)") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63162 "Found 2 new API test failures: TestWebKitAPI.ResourceLoadDelegate.ResourceType, TestWebKitAPI.PasteImage.RevealSelectionAfterPastingImage (failure)") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38096 "Found 1 new API test failure: /WPE/TestConsoleMessage:/webkit/WebKitConsoleMessage/security-error (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44374 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20644 "Found 8 new test failures: fast/events/onerror-img-after-gc.html http/tests/images/loading-image-no-border.html imported/w3c/web-platform-tests/html/canvas/element/manual/drawing-images-to-the-canvas/drawimage_crossorigin.sub.html imported/w3c/web-platform-tests/html/dom/reflection-embedded.html imported/w3c/web-platform-tests/html/semantics/embedded-content/the-img-element/attribute-order.html inspector/dom-debugger/url-breakpoints-dom-all-requests.html inspector/dom-debugger/url-breakpoints-dom-containing.html inspector/dom-debugger/url-breakpoints-dom-matching.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22983 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66189 "2 api tests failed or timed out") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20997 "Found 9 new test failures: fast/events/onerror-img-after-gc.html fast/picture/viewport-resize.html http/tests/media/media-blocked-by-willsendrequest.html imported/w3c/web-platform-tests/html/canvas/element/manual/drawing-images-to-the-canvas/drawimage_crossorigin.sub.html imported/w3c/web-platform-tests/html/dom/reflection-embedded.html imported/w3c/web-platform-tests/html/semantics/embedded-content/the-img-element/list-of-available-images-matching.https.html inspector/dom-debugger/url-breakpoints-dom-all-requests.html inspector/dom-debugger/url-breakpoints-dom-containing.html inspector/dom-debugger/url-breakpoints-dom-matching.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79297 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/730 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/230 "Found 8 new test failures: fast/events/onerror-img-after-gc.html http/tests/media/media-blocked-by-willsendrequest.html http/tests/paymentrequest/payment-response-payerName-attribute.https.html imported/w3c/web-platform-tests/html/canvas/element/manual/drawing-images-to-the-canvas/drawimage_crossorigin.sub.html imported/w3c/web-platform-tests/html/semantics/embedded-content/the-img-element/list-of-available-images-matching.https.html inspector/dom-debugger/url-breakpoints-dom-all-requests.html inspector/dom-debugger/url-breakpoints-dom-containing.html inspector/dom-debugger/url-breakpoints-dom-matching.html (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66094 "Found 7 new test failures: fast/events/onerror-img-after-gc.html http/tests/media/media-blocked-by-willsendrequest.html imported/w3c/web-platform-tests/html/semantics/embedded-content/the-img-element/list-of-available-images-matching.https.html inspector/dom-debugger/url-breakpoints-dom-all-requests.html inspector/dom-debugger/url-breakpoints-dom-containing.html inspector/dom-debugger/url-breakpoints-dom-matching.html security/block-test.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/871 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63174 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65373 "Found 1 new API test failure: /WebKitGTK/TestConsoleMessage:/webkit/WebKitConsoleMessage/security-error (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9214 "Passed tests") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7395 "Found 9 new test failures: fast/events/onerror-img-after-gc.html fast/picture/viewport-resize.html http/tests/media/media-blocked-by-willsendrequest.html imported/w3c/web-platform-tests/html/canvas/element/manual/drawing-images-to-the-canvas/drawimage_crossorigin.sub.html imported/w3c/web-platform-tests/html/dom/reflection-embedded.html imported/w3c/web-platform-tests/html/semantics/embedded-content/the-img-element/list-of-available-images-matching.https.html inspector/dom-debugger/url-breakpoints-dom-all-requests.html inspector/dom-debugger/url-breakpoints-dom-containing.html inspector/dom-debugger/url-breakpoints-dom-matching.html (failure)") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/693 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3430 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/723 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/707 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/725 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1382c0d807cf9583aa6ca231ce6aae74c8473349. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/31055

esprehn commented:
I'm pretty sure this PR queues a task not a micro task,  your test waits 500ms in a task to resolve the promise so it's of course after the task.

I think you need to follow the pattern I mentioned on Twitter to verify it's a microtask where you alter the values between the microtask and verify what gets fetched.

https://github.com/WebKit/WebKit/blob/180bfb4f1a6ef44cc14e4c8fdabfd134a60a5a42/Source/WebCore/dom/EventLoop.cpp#L183

esprehn commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -349,7 +349,10 @@ void ImageLoader::didUpdateCachedImage(RelevantMutation relevantMutation, Cached
     clearFailedLoadURL();
-    updateFromElement(relevantMutation);
+    Ref document = element().document();
+    document->eventLoop().queueTask(TaskSource::InternalAsyncTask, [imageLoader = this, relevantMutation] {
```
This should probably be queueMicrotask()?

yoavweiss replied:
yeah, for sure!

esprehn replied:
I have to read through the associated element code but is there denouncing somewhere?

yoavweiss replied:
What do you mean by debouncing in this context?

esprehn replied:
What I mean is that this method is called repeatedly each time an attribute changes:
https://github.com/WebKit/WebKit/blob/ed73610d6dc93d5e5b5065b34dabd11c878dcbf0/Source/WebCore/html/HTMLImageElement.cpp#L412

which means if you do:

img.src = "..."
img.referrerPolicy = "..."
img.crossOrigin = "..."

your code makes it queue 3 microtasks.

Maybe that's fine since the old code always ran updateFromElement() 3 times in that case too, but I think the spec intends only a single microtask in those situations.

yoavweiss replied:
I believe this is no longer the case. PTAL?

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ed73610d6dc93d5e5b5065b34dabd11c878dcbf0. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/31055

yoavweiss commented:
> I'm pretty sure this PR queues a task not a micro task, your test waits 500ms in a task to resolve the promise so it's after the task.
> 
> I think you need to follow the pattern I mentioned on Twitter to verify it's a microtask where you alter the values between the microtask and verify what gets fetched.

Done!

MentalGear commented:
I would greatly appreciate it if the srcset bug, which can cause unnecessary duplicate network requests, could finally be addressed following the amazing work by @yoavweiss  and review by @esprehn .

yoavweiss commented:
I need to get back to this (off hours hack..). Upcoming TPAC flights may be a good opportunity.

esprehn commented:
I have my review fingers ready!

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d290299a3f1881c6e50aa25417f1043a430d01dc. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/31055

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f1a1187a8e0e32060094b365a8c6337b7f60ed96. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/31055

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6808c774d5df40b4181bdda5bc0bcd4a0fcd33c9. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/31055

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d0756d67b315d6d5341a24ce3baa85410886e157. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/31055

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/91cceb8534b9821035e2f0b38f30350ba6a60163. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/31055

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/efbcca32e007262a568012bd71c16532a0b1d524. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/31055

yoavweiss commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -200,106 +204,154 @@ void ImageLoader::updateFromElement(RelevantMutation relevantMutation)
+    if (!document->getCodePosition()) {
+        auto codePosition = ContentSecurityPolicy::getCurrentCodePosition();
+        if (codePosition)
+            document->setCodePosition(codePosition.value());
```
Added a comment

shallawa replied:
It is still not clear to me why do we need to store the `getCurrentCodePosition()` in the document. 

This is a little weird pattern. I think you are storing the `getCurrentCodePosition()` in the document here to get it in `ContentSecurityPolicy::reportViolation()`. `Document` should not be aware of this `CodePosition` business.

yoavweiss commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -200,106 +204,154 @@ void ImageLoader::updateFromElement(RelevantMutation relevantMutation)
+        if (codePosition)
+            document->setCodePosition(codePosition.value());
+    }
+    // Grab the source location here, from the stack. Then store it on the document and exr=tract it on the other end
```
done!

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0214ddfc81f54b804e92ee31f76f6c6bd289b05b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/31055

esprehn commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -200,106 +204,155 @@ void ImageLoader::updateFromElement(RelevantMutation relevantMutation)
-        options.sameOriginDataURLFlag = SameOriginDataURLFlag::Set;
-        options.serviceWorkersMode = is<HTMLPlugInElement>(element()) ? ServiceWorkersMode::None : ServiceWorkersMode::All;
-        RefPtr imageElement = dynamicDowncast<HTMLImageElement>(element());
+    if (attr.isNull() || StringView(attr).containsOnly<isASCIIWhitespace<UChar>>()) {
```
I think you could actually change this to containsOnlyWhitespace() since that's apparently ASCII:
https://sourcegraph.com/github.com/ychin/EAWebKit/-/blob/EAWebKitSupportPackages/WTF/local/wtf/text/StringImpl.cpp?L314

shallawa replied:
I do not `containsOnlyWhitespace()` in our `StringImpl`. Are you suggesting adding it first and then calling it here?

esprehn reviewed (commented):
I think this looks correct now, though I don't have Reviewer powers on WebKit anymore.

LGTM

yoavweiss committed:
Fix srcset attribute setting order issues by aligning with spec

https://bugs.webkit.org/show_bug.cgi?id=190031
https://bugs.webkit.org/show_bug.cgi?id=276586

Reviewed by NOBODY (OOPS!).

Currently the order of setting image attributes matter, as some
attribute setting processing model triggers an image load.
Therefore setting some of these before others can impact the asset being
downloaded (or if an asset is downloaded at all).

As specified, this should not happen, as image loading awaits
a microtask before being triggered.

This PR aligns the implementation with the spec, and fixes these related
issues.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-img-element/attribute-order.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-img-element/attribute-order-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-img-element/list-of-available-images-matching.https-expected.txt: Amended.
* Source/WebCore/dom/Document.h: Add an m_codePosition member that can be
  used to carry the code position across the microtask.
 (WebCore::Document::getCodePosition):
 (WebCore::Document::setCodePosition):
* Source/WebCore/loader/ImageLoader.cpp:
 (WebCore::ImageLoader::doUpdateFromElement): Move some of the image
 loading logic to enable async loading.
 (WebCore::ImageLoader::updateFromElement): Remove the image loading
 logic that moved to doUpdateFromElement.
 (WebCore::ImageLoader::updateFromElementIgnoringPreviousError): Await
 a microtask before calling updateFromElement.
 (WebCore::ImageLoader::clearMicrotaskQueuedBool): clear the boolean
 indicating that a microtask was already queued.
* Source/WebCore/loader/ImageLoader.h: Add doUpdateFromElement.
* Source/WebCore/loader/cache/CachedResourceRequest.h: Make
  CachedResourceRequest fast-allocatable.
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
 (WebCore::ContentSecurityPolicy::getCurrentCodePosition): Move the
 code position retrival code to this helper.
 (WebCore::ContentSecurityPolicy::reportViolation const): Look at the
 document's codePosition (if there) before inspecting the stack.
* Source/WebCore/page/csp/ContentSecurityPolicy.h: Add a CodePosition
  struct.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/68b54ebce773ef4dffd64b9cb58f4370848c32c6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/31055

shallawa commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -206,106 +211,160 @@ void ImageLoader::updateFromElement(RelevantMutation relevantMutation)
-        options.serviceWorkersMode = is<HTMLPlugInElement>(element()) ? ServiceWorkersMode::None : ServiceWorkersMode::All;
-        RefPtr imageElement = dynamicDowncast<HTMLImageElement>(element());
+    if (attr.isNull() || StringView(attr).containsOnly<isASCIIWhitespace<UChar>>()) {
+        if (!attr.isNull()) {
+            m_failedLoadURL = attr;
+            m_hasPendingErrorEvent = true;
+            loadEventSender().dispatchEventSoon(*this, eventNames().errorEvent);
+        }
+        didUpdateCachedImage(relevantMutation, WTFMove(newImage));
+        return;
+    }
```
I think making this early return is a clean-up step. Can we start by making it in a separate PR? Also please let make things clear even if they are verbose a little:
```
if (StringView(attr).containsOnly<isASCIIWhitespace<UChar>>()) {
    ...
    didUpdateCachedImage(relevantMutation, WTFMove(newImage));
    return;
}

if (!attr.isNull()) {
    didUpdateCachedImage(relevantMutation, WTFMove(newImage));
    return;
}
```

shallawa commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -206,106 +211,160 @@ void ImageLoader::updateFromElement(RelevantMutation relevantMutation)
+            // if "src" attributeChanged is not called yet, imageURL can be null and it does not work well for ResourceRequest.
+            // In this case, we should behave same as attr.isNull().
+            imageURL = imageElement->currentURL();
+            if (!imageURL.isValid()) {
```
This used to be 
```
if (imageURL.isNull()) {
```
And the comment above is saying:
```
if "src" attributeChanged is not called yet, imageURL can be null
```
Can you please explain why did you make this change and fix the comment if it is really needed?

shallawa commented on Source/WebCore/loader/ImageLoader.h:
```diff
@@ -147,6 +152,7 @@ class ImageLoader : public CachedImageClient {
     bool m_imageComplete : 1;
     bool m_loadManually : 1;
     bool m_elementIsProtected : 1;
+    bool m_microtaskQueued : 1;
```
I would suggest renaming this flag `m_isMicrotaskQueued` and the setter and getter to be 
```
bool isMicrotaskQueued() const
void setIsMicrotaskQueued(bool isMicrotaskQueued)
```

shallawa commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -206,106 +211,160 @@ void ImageLoader::updateFromElement(RelevantMutation relevantMutation)
+
+    auto request = createPotentialAccessControlRequest(WTFMove(resourceRequest), WTFMove(options), document, crossOriginAttribute);
+    request.setInitiator(element());
+    m_imageComplete = false;
```
This is new. What is going to set it to true?

shallawa commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -206,106 +211,160 @@ void ImageLoader::updateFromElement(RelevantMutation relevantMutation)
+    if (imageElement) {
+        URL imageURL = imageElement->currentURL();
+        if (request.resourceRequest().url() != imageURL) {
+            ResourceLoaderOptions options = CachedResourceLoader::defaultCachedResourceOptions();
+            options.contentSecurityPolicyImposition = protectedElement()->isInUserAgentShadowTree() ? ContentSecurityPolicyImposition::SkipPolicyCheck : ContentSecurityPolicyImposition::DoPolicyCheck;
+            options.loadedFromPluginElement = is<HTMLPlugInElement>(element()) ? LoadedFromPluginElement::Yes : LoadedFromPluginElement::No;
+            options.sameOriginDataURLFlag = SameOriginDataURLFlag::Set;
+            options.serviceWorkersMode = is<HTMLPlugInElement>(element()) ? ServiceWorkersMode::None : ServiceWorkersMode::All;
```
This is part is repeated in `updateFromElement()`. Can we remove this duplication?

shallawa reviewed (commented):
Can you please submit a PR for the `CodePosition` and another PR for cleaning `updateFromElement()` and then rebase this PR? This should make reviewing this PR and investigating the layout failures a lot easier.
